pkgs/test_core/lib/test_core.dart

```diff
@@ -285,3 +286,33 @@ void setUpAll(dynamic Function() callback) => _declarer.setUpAll(callback);
 /// prohibitively slow.
 void tearDownAll(dynamic Function() callback) =>
     _declarer.tearDownAll(callback);
+
+/// Base class that may be used by a test to create a fake object that should
+/// never be used.  For example:
+///
+///     class FakeFoo extends Fake implements Foo {}
+///
+/// The behavior of a fake object is similar to `null`, except that:
+/// - It may be passed into a non-nullable API.
+/// - It satisfies `is` checks for the type it is mocking.
+/// - It throws an exception for any method invocation, whereas `null` supports
+///   toString(), hashCode, etc.
+class Fake {
```

Comment:
If we decide to ship this it should be in `test_api`, and the tests should go there too.